code/game/objects/structures/tables_racks.dm

```diff
@@ -784,11 +784,31 @@
 			continue
 		held.forceMove(NewLoc)
 
+/obj/structure/table/tray/can_be_pulled(user, grab_state, force, show_message)
+	var/atom/movable/held
+	for(var/held_uid in held_items)
+		held = locateUID(held_uid)
+		if(!held)
+			held_items -= held_uid
+			continue
+		if(loc != held.loc)
+			held_items -= held_uid
+			continue
```

Comment:
Since the held items are only updated on move(), it means that if you climb on top of the tray, then try to pull it you will find yourself unable to do so (Since your UID is still in held_items), until the move the tray, though I should probably move it to a separate proc so it's not just duplicate code.